### ISSUE TITLE
ResourceSet Level Generation

### ISSUE THREAD

**Issue by jGleitz** [eclipse/xtext-core#413]
Xtext’s Generator framework seems to be designed for situations where each input file creates one ore more output files (which is arguably the common case). 

If, however, several input files contribute to one output file, the way `IGenerator`(`2`) works is problematic, because there is no way (that I would know of) to be called once per build with all changed resources.

My current strategy is to retrieve all input files from the resource set the current input file is part of and generate the output file from them. Then store the resource set’s hash in a static field and only regenerate when seeing a new resource set (determined by that hash). That strategy is hacky and my first question is whether it’s valid, i.e. if 

 * all resources of a Xtext-DSL will be in the same resource set in an Xtext build
 * every build will use a new resource set

What do you think about adding an API for Generators that want do be called once per build; to support the use case of creating one output file out of multiple input files? Or is there already something like that?

**Comment by cdietrich** [comment]
The basic principle is described here

https://kthoms.wordpress.com/2011/07/12/xtend-generating-from-multiple-input-models/

Some hints how to port this to current apis here
https://www.eclipse.org/forums/index.php?t=msg&th=1087436&goto=1767975&

**Comment by jGleitz** [comment]
@cdietrich Thanks a lot for those hints.

But as `BuilderParticipant` is an Eclipse-specific class, there is no platform-independent way to achieve this, is there? (E.g. something that would also work when building with Maven).

**Comment by cdietrich** [comment]
Hmm that would be not possible without forking the maven plugin

**Comment by jGleitz** [comment]
So would it be possible to add an interface for this to Xtext? Because then `BuilderParticipant` and the Maven Plugin could be adapted accordingly. 

I think the use case is fairly common and the amount of required changes comparably small.

**Comment by cdietrich** [comment]
Well I would have to have a look at the API the maven and the gradle plugin use.

**Comment by cdietrich** [comment]
Have a look at the discussion in https://bugs.eclipse.org/bugs/show_bug.cgi?id=351903 as well

**Comment by jGleitz** [comment]
> Have a look at the discussion in https://bugs.eclipse.org/bugs/show_bug.cgi?id=351903 as well

Yes, this discusses the same issue (and proves that there is a need for it).

The solution proposed there is the same I’m proposing, only that the new Interface would be provided with deltas. This surely is a more advanced solution, but it also requires more work (which is propably why nobody has done it yet.)

In my use case, I have a decent strategy to update existing files. So just getting the resources that changed (and not their deltas) would be sufficient. It would be a lot easier to implement, though, because there is no need to [synthesize deltas](https://bugs.eclipse.org/bugs/show_bug.cgi?id=351903#c9) for standalone builds (those would simply give all resources to the new Interface).

P.S.: What’s your policy for old bugs: should we move the discussion to the Eclipse Bug or post a link to this issue there?

**Comment by kthoms** [comment]
For a standalone generator the approach is: Extend the GeneratorComponent class and call the extended generator implementation using the resources of the slots by extending invoke(). Note that the generator is injected with GeneratorDelegate, which also has to be extended. Then you should be able to follow my approach for a CLI generator.

**Comment by kthoms** [comment]
> P.S.: What’s your policy for old bugs: should we move the discussion to the Eclipse Bug or post a link to this issue there?
We prefer discussion here now, but we are also fine with Bugzilla.

**Comment by svenefftinge** [comment]
As mentioned in teh related discussions, the main issue is that you cannot eb sure that the resourceset is 'complete', as Xtext builders support incremental updates.

If you can somehow make sure that th resourceset is complete, then it is very simple to make sure you generate only once. 
For instance, you could trigger generation only for the first resource in the resourceset's resources list.

**Comment by jGleitz** [comment]
@svenefftinge Although I’ve read the discussion, I’m not quite sure what you mean by “you cannot be sure that the resourceset is 'complete'”.

In said discussion, you wrote 

> My proposal for some sort of middle ground (and yet correct) solution is to introduce a callback interface, which gets all deltas, the information whether it's a full build or incremental build and the resourceSet.

If I see the situation right, this hasn’t been realised yet because of the high effort needed to get platform independent deltas. So I propose a more simplified version: An interface that only gets the resource set and the info whether it’s an incremental build. This would already heavily improve my use case.

**Comment by cdietrich** [comment]
there is already a interface for that: IXtextBuilderParticipant

**Comment by jGleitz** [comment]
I just realised I never replied back.

@cdietrich as I already explained above, `IXtextBuilderParticipant` does not fit my use case because I need something that will also work for Maven.

I still think an interface for ResourceSet-level generation, that works with Maven and Eclipse, would be very beneficial, even without deltas.

**Comment by cdietrich** [comment]
i fear there is nothing about besides generate for each occurrent and override the generated file

**Comment by cdietrich** [comment]
will close this one as wontfix. feel free to reopen in github.com/eclipse/xtext in case known workaround dont fit
